Handle error events in HCS notifications.

helsaawy opened the pull request:
The shim parses the JSON result document and handles the error events (via `processHcsResult`) returned directly by HCS calls (e.g.,`vmcompute.HcsCreateComputeSystem`), but ignores the JSON (data) payload for notifications (which are either received from a  `processAsyncHcsResult` in the appropriate system or process call in `"internal.hcs"`, or via `waitForNotification` in `waitBackground`).

This leads to ambiguous failure errors (e.g., `"The data is invalid."`) that require ETW traces to track down the appropriate HCS logs, even though the error events already provide enough information to identify the issue.

Add `processNotification()` to:
    1. parse the `notificationData` JSON payload provided by HCS to the `notificationWatcher` callback into the an `hcsResult` struct; and
    2. validate that the JSON data matches the notification HResult.

Create a new error type (`resultError`) to mirror the [HCS ResultError][result-error] and consolidate the events alongside the original error.

Also, since `ErrorEvent`s are always converted to strings in the context of concatenating several of them together, add an
`(*ErrorEvent).writeTo(*string.Builder)` function to provide more efficient error string generation for `(HCS|System|Process)Error`s. 
Additionally, consolidate the joining and formatting of error events for those error types.

Requires: #2602

[result-error]: https://learn.microsoft.com/en-us/virtualization/api/hcs/schemareference#ResultError

helsaawy commented:
>Should we have an effort to move shim from HCS V1 to V2?

Ideally, yes, we should, but I am not sure the work involved would be easily addressed in a single PR, so this PR should be a good stopgap until we move entirely over to the computecore dlls

rawahars commented on internal/hcs/errors.go:
```diff
@@ -99,48 +103,85 @@ type ErrorEvent struct {
 	}
 	if ev.EventID != 0 {
-		evs = fmt.Sprintf("%s EventID: %d", evs, ev.EventID)
+		fmt.Fprintf(b, " EventID: %d", ev.EventID)
```
We should perhaps use the same semantics for writing i.e. `WriteString`.

helsaawy replied:
wed have to call `b.WriteString(fmt.Sprintf(" EventID: %d", ev.EventID))` to handle the formatting, which feels more convoluted than `Fprintf`

rawahars commented on internal/hcs/errors.go:
```diff
@@ -99,48 +103,85 @@ type ErrorEvent struct {
+	_, _ = b.WriteString("[Event Detail: " + ev.Message)
 	if ev.StackTrace != "" {
-		evs += " Stack Trace: " + ev.StackTrace
+		_, _ = b.WriteString(" Stack Trace: " + ev.StackTrace)
```
Should we have a mechanism to handle the error in the builder method?

rawahars commented on internal/hcs/errors.go:
```diff
@@ -99,48 +103,85 @@ type ErrorEvent struct {
 	}
-	evs += "]"
-	return evs
+	_ = b.WriteByte(']')
```
Why are we using `WriteByte` instead of `WriteString`?

helsaawy committed:
Minor code cleanup and doc fixes.

Clarify docs.
Fix spelling.
Remove newline from comments.
Don't serialize `OptionsWCOW.OutputHandlerCreator` to JSON.
Consolidate redundant `internal/winapi/winapi.go` with `internal/winapi/doc.go`.

Use `"internal/log".(*Hook)` to format logs and add span information
in functional/gcs tests.
Add testing `nil`-checks and logs to uVM/container/process testing
operations.

Signed-off-by: Hamza El-Saawy <hamzaelsaawy@microsoft.com>

helsaawy committed:
Invalid Handle check; log field names

Add `"internal/winapi".IsInvalidHandle` function to valid `Handle`s are
neither `0` nor Invalid.

Standardize system and process ID log/span field names.
Use `"system-id"` for system ID rather than `"cid"`, which can be
interpreted as "container ID".

Add `(*Process).CloseCtx(context.Context)` and use that, along with
`(*System).CloseCtx` where appropriate.

Suppress lint issues (skipped error checks for `Close()` calls
are ignored automatically, but not for `CloseCtx(ctx)`).

Update CI lint version.

Signed-off-by: Hamza El-Saawy <hamzaelsaawy@microsoft.com>

helsaawy committed:
Use dedicated callback number type and atomic counter

Add an explicit `callbackNumber` type and use an `atomic.Uintptr` to
track callback numbers instead of the RWMutex `callbackMapLock`.

Add comment clarifying use of `callbackNumber` as callback context.

Add logs to `notificationWatcher` and `(*Process|*System)[un]registryCallback`.

Signed-off-by: Hamza El-Saawy <hamzaelsaawy@microsoft.com>

helsaawy committed:
Handle error events in HCS notifications.

The shim parses the JSON result document and handles the error events
(via `processHcsResult`) returned directly by HCS calls
(e.g.,`vmcompute.HcsCreateComputeSystem`), but ignores the JSON (data)
payload for notifications (which are either received from a
`processAsyncHcsResult` in the appropriate system or process call in
`"internal.hcs"`, or via `waitForNotification` in `waitBackground`).

This leads to ambiguous failure errors (e.g., `"The data is invalid."`)
that require ETW traces to track down the appropriate HCS logs, even
though the error events already provide enough information to identify
the issue.

Add `processNotification()` to:
 1. parse the `notificationData` JSON payload provided by HCS to the
    `notificationWatcher` callback into the an `hcsResult` struct; and
 2. validate that the JSON data matches the notification HResult.

Create a new error type (`resultError`) to mirror the original [HCS
ResultError][result-error] and consolidate the events alongside the
original error.

Also, since `ErrorEvent`s are always converted to strings in the
context of concatenating several of them together, add an
`(*ErrorEvent).writeTo(*string.Builder)` function to provide more
efficient error string generation for `(HCS|System|Process)Error`s.
Additionally, consolidate the joining and formatting of error events
for those error types.

[result-error]: https://learn.microsoft.com/en-us/virtualization/api/hcs/schemareference#ResultError

Signed-off-by: Hamza El-Saawy <hamzaelsaawy@microsoft.com>
